File: scrapers/torrent_info.py
import re
from urllib.parse import urljoin

from scrapers.helpers import get_page_bs4
from utils.torrent import get_info_hash_from_magnet


async def get_torrent_info(url: str, indexer: str) -> dict:
    torrent_info = {}
    pre_process_func = pre_process_map.get(indexer, get_page_bs4)
    parse_func = parser_function_map.get(indexer, parse_common_torrents)

    soup = await pre_process_func(url)
    if not soup:
        return torrent_info

    torrent_info = parse_func(soup, torrent_info, url)
    return torrent_info


def parse_1337x(soup, torrent_info, url):
    magnet_link = soup.find("a", href=re.compile("magnet:?"))
    if magnet_link:
        torrent_info["magnetUrl"] = magnet_link.get("href")
        torrent_info["downloadUrl"] = None

    info_hash = soup.find("div", class_="infohash-box").find("span")
    if info_hash:
        torrent_info["infoHash"] = info_hash.text.strip()

    return torrent_info


def parse_torrent_downloads(soup, torrent_info, url):
    magnet_link = soup.find("a", href=re.compile("magnet:?"))
    if magnet_link:
        torrent_info["magnetUrl"] = magnet_link.get("href")
        torrent_info["downloadUrl"] = None

    info_hash_span = soup.find("div", id="main_wrapper").find("span", string=re.compile("Infohash:"))
    if info_hash_span:
        info_hash = info_hash_span.find_parent("p").text.replace("Infohash:", "").strip()
        torrent_info["infoHash"] = info_hash

    return torrent_info


def parse_torrentdownloads(soup, torrent_info, url):
    info_hash = soup.find("td", string=re.compile("[A-Fa-f0-9]{40}"))
    if info_hash:
        torrent_info["infoHash"] = info_hash.text.strip()

    torrent_download_link = soup.find("a", href=re.compile("/td.php?"))
    magnet_link = soup.find("a", href=re.compile("magnet:?"))
    if torrent_download_link:
        torrent_info["downloadUrl"] = urljoin(url, torrent_download_link.get("href"))
        torrent_info["magnetUrl"] = None
    elif magnet_link:
        torrent_info["magnetUrl"] = magnet_link.get("href")
        torrent_info["downloadUrl"] = None

    return torrent_info


def parse_badass_torrents(soup, torrent_info, url):
    info_hash = soup.find("td", string=re.compile("[A-Fa-f0-9]{40}"))
    if info_hash:
        torrent_info["infoHash"] = info_hash.text.strip()

    magnet_link = soup.find("a", href=re.compile("magnet:?"))
    torrent_link = soup.find("a", string="Torrent Download")
    if torrent_link:
        torrent_info["downloadUrl"] = urljoin(url, torrent_link.get("href"))
    if magnet_link:
        torrent_info["magnetUrl"] = magnet_link.get("href")
    return torrent_info


def parse_common_torrents(soup, torrent_info, url):
    magnet_link = soup.find("a", href=re.compile("magnet:?"))
    if magnet_link:
        torrent_info["magnetUrl"] = magnet_link.get("href")
        torrent_info["downloadUrl"] = None
        torrent_info["infoHash"] = get_info_hash_from_magnet(magnet_link.get("href"))
    return torrent_info


def parse_itorrent(soup, torrent_info, url):
    magnet_link = soup.find("a", href=re.compile("magnet:?"))
    if magnet_link:
        torrent_info["infoHash"] = get_info_hash_from_magnet(magnet_link.get("href"))

    download_link = soup.find("a", class_="jq-download")
    if download_link:
        torrent_info["magnetUrl"] = None
        torrent_info["downloadUrl"] = urljoin(url, download_link.get("href"))

    return torrent_info


def parse_gktorrent(soup, torrent_info, url):
    magnet_link = soup.find("a", href=re.compile("magnet:?"))
    if magnet_link:
        torrent_info["magnetUrl"] = None
        torrent_info["infoHash"] = get_info_hash_from_magnet(magnet_link.get("href"))

    download_link = soup.find("div", class_="btn-download").find("a")
    if download_link:
        torrent_info["downloadUrl"] = urljoin(url, download_link.get("href"))

    return torrent_info


async def pre_process_therarbg(url):
    url = url.replace("?format=json", "")
    return await get_page_bs4(url)


pre_process_map = {
    "TheRARBG": pre_process_therarbg,
}

parser_function_map = {
    "1337x": parse_1337x,
    "TheRARBG": parse_common_torrents,
    "Torrent Downloads": parse_torrent_downloads,
    "Badass Torrents": parse_badass_torrents,
    "iTorrent": parse_itorrent,
    "TorrentDownloads": parse_torrentdownloads,
}

File: backend/src/scrapers/torrent_info.rs
//! Fetch magnet/download URLs from public indexer detail pages.
//!
//! Mirrors Python `workers/scrapers/torrent_info.py`.

use std::collections::HashMap;
use std::sync::OnceLock;
use std::time::Duration;

use regex::Regex;
use reqwest::Client;
use scraper::{Html, Selector};

use crate::parser;

#[derive(Debug, Default, Clone)]
pub struct TorrentInfoPage {
    pub magnet_url: Option<String>,
    pub download_url: Option<String>,
    pub info_hash: Option<String>,
}

pub async fn get_torrent_info(
    http: &Client,
    url: &str,
    indexer: &str,
    timeout: Duration,
) -> TorrentInfoPage {
    let fetch_url = pre_process_url(indexer, url);
    let html = match http.get(&fetch_url).timeout(timeout).send().await {
        Ok(resp) if resp.status().is_success() => resp.text().await.unwrap_or_default(),
        Ok(resp) => {
            tracing::debug!("torrent_info: HTTP {} for {fetch_url}", resp.status());
            return TorrentInfoPage::default();
        }
        Err(e) => {
            tracing::debug!(
                error_kind = crate::util::http::transport_error_kind(&e),
                "torrent_info: fetch failed for {fetch_url}: {e}"
            );
            return TorrentInfoPage::default();
        }
    };

    parse_detail_page(indexer, &html, &fetch_url)
}

fn pre_process_url(indexer: &str, url: &str) -> String {
    if indexer.eq_ignore_ascii_case("TheRARBG") {
        url.replace("?format=json", "")
    } else {
        url.to_string()
    }
}

fn parser_for(indexer: &str) -> fn(&Html, &str) -> TorrentInfoPage {
    parsers()
        .get(indexer)
        .copied()
        .unwrap_or(parse_common_torrents)
}

fn parse_detail_page(indexer: &str, html: &str, url: &str) -> TorrentInfoPage {
    let doc = Html::parse_document(html);
    parser_for(indexer)(&doc, url)
}

fn parse_1337x(doc: &Html, _url: &str) -> TorrentInfoPage {
    let mut info = TorrentInfoPage::default();
    if let Some(magnet) = first_href(doc, "a[href^='magnet:?']") {
        info.magnet_url = Some(magnet);
    }
    if let Some(hash) = doc
        .select(&Selector::parse("div.infohash-box span").expect("infohash-box span"))
        .next()
        .map(|el| el.text().collect::<String>().trim().to_string())
        .filter(|h| h.len() == 40)
    {
        info.info_hash = Some(hash);
    }
    info
}

fn parse_torrent_downloads(doc: &Html, _url: &str) -> TorrentInfoPage {
    let mut info = TorrentInfoPage::default();
    if let Some(magnet) = first_href(doc, "a[href^='magnet:?']") {
        info.magnet_url = Some(magnet);
    }
    if let Some(wrapper) = doc
        .select(&Selector::parse("div#main_wrapper").expect("main_wrapper"))
        .next()
    {
        let wrapper_text = wrapper.text().collect::<String>();
        if let Some(re) = INFO_HASH_RE.get()
            && let Some(m) = re.find(&wrapper_text)
        {
            info.info_hash = Some(m.as_str().to_string());
        }
    }
    info
}

fn parse_torrentdownloads(doc: &Html, url: &str) -> TorrentInfoPage {
    let mut info = TorrentInfoPage::default();
    if let Some(hash) = doc
        .select(&Selector::parse("td").expect("td"))
        .find(|el| {
            let text = el.text().collect::<String>();
            INFO_HASH_RE
                .get()
                .is_some_and(|re| re.is_match(text.trim()))
        })
        .map(|el| el.text().collect::<String>().trim().to_string())
    {
        info.info_hash = Some(hash);
    }
    if let Some(href) = first_href(doc, "a[href*='/td.php?']") {
        info.download_url = Some(resolve_url(url, &href));
    } else if let Some(magnet) = first_href(doc, "a[href^='magnet:?']") {
        info.magnet_url = Some(magnet);
    }
    info
}

fn parse_badass_torrents(doc: &Html, url: &str) -> TorrentInfoPage {
    let mut info = TorrentInfoPage::default();
    if let Some(hash) = doc
        .select(&Selector::parse("td").expect("td"))
        .find(|el| {
            let text = el.text().collect::<String>();
            INFO_HASH_RE
                .get()
                .is_some_and(|re| re.is_match(text.trim()))
        })
        .map(|el| el.text().collect::<String>().trim().to_string())
    {
        info.info_hash = Some(hash);
    }
    if let Some(magnet) = first_href(doc, "a[href^='magnet:?']") {
        info.magnet_url = Some(magnet);
    }
    for el in doc.select(&Selector::parse("a").expect("a")) {
        let text = el.text().collect::<String>();
        if text.trim() == "Torrent Download" {
            if let Some(href) = el.value().attr("href") {
                info.download_url = Some(resolve_url(url, href));
            }
            break;
        }
    }
    info
}

fn parse_itorrent(doc: &Html, url: &str) -> TorrentInfoPage {
    let mut info = TorrentInfoPage::default();
    if let Some(magnet) = first_href(doc, "a[href^='magnet:?']") {
        info.info_hash = parser::extract_info_hash(&magnet).map(|h| h.to_string());
    }
    if let Some(href) = first_href(doc, "a.jq-download") {
        info.download_url = Some(resolve_url(url, &href));
    }
    info
}

fn parse_gktorrent(doc: &Html, url: &str) -> TorrentInfoPage {
    let mut info = TorrentInfoPage::default();
    if let Some(magnet) = first_href(doc, "a[href^='magnet:?']") {
        info.info_hash = parser::extract_info_hash(&magnet).map(|h| h.to_string());
    }
    let download_sel = Selector::parse("div.btn-download a").expect("div.btn-download a");
    if let Some(href) = doc
        .select(&download_sel)
        .next()
        .and_then(|el| el.value().attr("href"))
    {
        info.download_url = Some(resolve_url(url, href));
    }
    info
}

fn parse_common_torrents(doc: &Html, _url: &str) -> TorrentInfoPage {
    let mut info = TorrentInfoPage::default();
    if let Some(magnet) = first_href(doc, "a[href^='magnet:?']") {
        info.magnet_url = Some(magnet.clone());
        info.info_hash = parser::extract_info_hash(&magnet).map(|h| h.to_string());
    }
    info
}

fn first_href(doc: &Html, selector: &str) -> Option<String> {
    let sel = Selector::parse(selector).ok()?;
    doc.select(&sel)
        .next()
        .and_then(|el| el.value().attr("href"))
        .map(|href| href.to_string())
}

fn resolve_url(base: &str, href: &str) -> String {
    if href.starts_with("http://") || href.starts_with("https://") {
        return href.to_string();
    }
    if let Ok(abs) = reqwest::Url::parse(base).and_then(|b| b.join(href)) {
        return abs.to_string();
    }
    href.to_string()
}

type TorrentInfoParser = fn(&Html, &str) -> TorrentInfoPage;

static INFO_HASH_RE: OnceLock<Regex> = OnceLock::new();
static PARSERS: OnceLock<HashMap<&'static str, TorrentInfoParser>> = OnceLock::new();

fn parsers() -> &'static HashMap<&'static str, TorrentInfoParser> {
    PARSERS.get_or_init(|| {
        HashMap::from([
            ("1337x", parse_1337x as TorrentInfoParser),
            ("TheRARBG", parse_common_torrents),
            ("Torrent Downloads", parse_torrent_downloads),
            ("Badass Torrents", parse_badass_torrents),
            ("iTorrent", parse_itorrent),
            ("GkTorrent", parse_gktorrent),
            ("TorrentDownloads", parse_torrentdownloads),
        ])
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registry_includes_requested_indexers() {
        let map = parsers();
        assert!(map.contains_key("Badass Torrents"));
        assert!(map.contains_key("iTorrent"));
        assert!(map.contains_key("GkTorrent"));
    }

    #[test]
    fn parse_common_extracts_magnet_hash() {
        let html = r#"<html><body><a href="magnet:?xt=urn:btih:0123456789abcdef0123456789abcdef01234567">mag</a></body></html>"#;
        let doc = Html::parse_document(html);
        let info = parse_common_torrents(&doc, "https://example.com");
        assert!(info.magnet_url.is_some());
        assert_eq!(
            info.info_hash.as_deref(),
            Some("0123456789abcdef0123456789abcdef01234567")
        );
    }
}
